docs: IFS-3028 revise Spring & Spring Boot guidelines

wfahmer-msg opened the pull request.

wfahmer-msg committed:
feat: IFS-3028 - fix Spring dependency name

wfahmer-msg committed:
feat: IFS-3028 - rework and shorten Spring introduction

wfahmer-msg committed:
feat: IFS-3028 - Spring core concepts

wfahmer-msg committed:
feat: IFS-3028 - remove webAppRootKey configuration

wfahmer-msg committed:
feat: IFS-3028 - Configure SpringBoot

wfahmer-msg committed:
feat: IFS-3028 - Configure SpringBootApplication

wfahmer-msg committed:
feat: IFS-3028 - adjust image include

wfahmer-msg committed:
feat: IFS-3028 - adjust image include

wfahmer-msg committed:
feat: IFS-3028 - adjust image include

wfahmer-msg committed:
feat: IFS-3028 - SpringBoot @Component

wfahmer-msg committed:
feat: IFS-3028 - Bean scopes

wfahmer-msg committed:
feat: IFS-3028 - rework AOP

wfahmer-msg committed:
feat: IFS-3028 - rework dependency injection

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/serviceschicht.adoc:
```diff
@@ -79,7 +79,7 @@ Dabei ist sicherzustellen, dass zum Zeitpunkt der Erstellung der Log-Einträge e
 Es wandelt die in der HTTP-Anfrage übertragenen Daten in Transportobjekte um und ruft die entsprechende Fachkomponente auf.
 Ebenso erledigt es die Umwandlung des Ergebnisses in eine HTTP-Antwort.
-Zur Definition der Service-Endpunkte wird Spring Web MVC verwendet.
+Zur Definition der Service-Endpunkte wird Spring Web verwendet.
```
Wir empfehlen Spring Web MVC: https://isyfact.github.io/isyfact-standards-doku/current/isy-service-rest/konzept/master.html#spring-mvc-vergleich-webflux

stdll replied:
Spring Web ist der Oberbegriff für sowohl den Servlet- als auch Reactive-Stack. Die IsyFact empfiehlt aktuell Spring MVC, sperrt sich aber auch nicht gegen die Nutzung von Spring Webflux.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -2,9 +2,12 @@
+Das Spring-Framework ist ein leichtgewichtiges Java Framework, welches sehr verständlich und leicht zu verwenden ist.
+Zu den Kernkonzepten von Spring gehören Inversion of Control (IoC) und Dependency Injection (DI).
+
+IoC ist ein Umsetzungsparadigma und bedeutet Steuerungsumkehr. Im Kontext des Spring Frameworks versteht man darunter, dass das Framework die Erstellung des Objektnetzes (_Spring-Beans_) anstelle des Entwicklers übernimmt.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -2,9 +2,12 @@
+
+IoC ist ein Umsetzungsparadigma und bedeutet Steuerungsumkehr. Im Kontext des Spring Frameworks versteht man darunter, dass das Framework die Erstellung des Objektnetzes (_Spring-Beans_) anstelle des Entwicklers übernimmt.
+
+_Spring-Beans_ sind Java Objekte, die durch den Spring IoC Container instanziiert und verwaltet werden. Der IoC Container erstellt _Spring-Beans_ anhand einer Bean Definition, die der Entwickler in Form von Annotationen oder xml Konfiguration bereitstellt.
```
Nur ein Satz pro Zeile.

dauna-msg replied:
Bitte noch einmal prüfen, ob wir eine feste Vorgabe für die Definition haben. Mir sind in der IF-Welt nur Annotationen und keine XML-Konfigurationen bekannt.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -13,68 +16,30 @@ Die Nutzung spezifischer Features von Spring sind Bestandteile der folgenden Bau
+Spring Boot vereinfacht die Konfiguration durch:
 
-[[beans-standardmaessig-als-singletons-definieren]]
-==== Beans standardmäßig als Singletons definieren
```
Hier fehlt mir die konkrete Vorgabe, dass Scope = request für Aufrufe durch HTTP-Requests verwendet werden soll und singleton für alles andere.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -13,68 +16,30 @@ Die Nutzung spezifischer Features von Spring sind Bestandteile der folgenden Bau
-Die Spring-Beans einer Anwendung sollen zustandslos sein und werden als Singleton-Beans erzeugt.
-Wo technisch erforderlich können auch andere Scopes verwendet werden.
-
-NOTE: Eine genaue Beschreibung der Scopes bietet die https://docs.spring.io/spring-framework/reference/core/beans/factory-scopes.html[offizielle Spring-Dokumentation].
```
Der Link ist doch hilfreich, würde ich beibehalten.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -13,68 +16,30 @@ Die Nutzung spezifischer Features von Spring sind Bestandteile der folgenden Bau
+Diese Annotation ist äquivalent zur gleichzeitigen Nutzung der Annotationen  '@Configuration', '@EnableAutoConfiguration' und '@ComponentScan'.
+
+==== Spring Boot SpringApplication Klasse
+Die Spring Boot SpringApplication-Klasse wird verwendet, um eine Spring-Anwendung aus einer Java-Hauptmethode zu booten und zu starten. Diese Klasse erstellt automatisch den https://docs.spring.io/spring-framework/reference/core/beans/context-introduction.html[ApplicationContext] aus dem Klassenpfad, scannt die Konfigurationsklassen und startet die Anwendung.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -83,46 +48,24 @@ Konfiguration mit `@SpringBootApplication`:
 
-Konfiguration mit `@Configuration` und `@EnableAutoConfiguration`:
+==== Spring Boot @Configuration Annotation
+Die Annotation Spring '@Configuration' zeigt an, dass die Klasse @Bean-Definitionsmethoden hat. So kann der Spring-Container die Klasse verarbeiten und Spring-Beans erzeugen, die in der Anwendung verwendet werden.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -83,46 +48,24 @@ Konfiguration mit `@SpringBootApplication`:
+Die Annotation Spring '@Configuration' zeigt an, dass die Klasse @Bean-Definitionsmethoden hat. So kann der Spring-Container die Klasse verarbeiten und Spring-Beans erzeugen, die in der Anwendung verwendet werden.
 
-[[listing-applikationsklasse]]
-.Applikationsklasse mit expliziter Konfiguration
```
Warum fällt dies raus?

stdll replied:
Es ist ein für Spring Boot sehr unüblicher Weg, Applikationen so zu konfigurieren.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+==== Spring Boot @Component Annotation
+@Component ist die generische Annotation, die verwendet wird, um eine Klasse als Spring-Bean zu markieren.
+Sie wird für jede Klasse verwendet, die als Spring-Bean verwaltet werden soll, unabhängig von ihrer Funktion oder Verantwortung.
+Es ist eine allgemeine Annotation und vermittelt keine spezifische Bedeutung über die Rolle oder Funktion der bean.
```
bean -> Bean

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
-[[webapprootkey-konfigurieren]]
-=== webAppRootKey konfigurieren
+==== Spring Boot @Service Annotation
+Die Annotation @Service ist eine spezielle Form von @Component. Es ist eine semantische Annotation, die verwendet wird, um Klassen zu kennzeichne die Geschäftsfunktionen implementieren.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+Die Annotation Spring Controller ist eine Spezialisierung der Annotation @Component. Die Spring Controller-Annotation wird typischerweise in Kombination mit annotierten Handler-Methoden verwendet, die auf der RequestMapping-Annotation basieren.
+
+===== Spring Controller
+Die Spring Controller-Annotation kann nur auf Klassen angewendet werden. Sie wird verwendet, um eine Klasse als Web-Request-Handler zu markieren. Sie wird meist bei Spring MVC-Anwendungen verwendet.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
-</context-param>
-----
+==== Spring Boot @Controller und @RestController Annotation
+Die Annotation Spring Controller ist eine Spezialisierung der Annotation @Component. Die Spring Controller-Annotation wird typischerweise in Kombination mit annotierten Handler-Methoden verwendet, die auf der RequestMapping-Annotation basieren.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+Die Spring Controller-Annotation kann nur auf Klassen angewendet werden. Sie wird verwendet, um eine Klasse als Web-Request-Handler zu markieren. Sie wird meist bei Spring MVC-Anwendungen verwendet.
+
+===== Spring RestController
+Spring @RestController ist eine Komfort-Annotation, die selbst mit @Controller und @ResponseBody annotiert ist. Diese Annotation wird verwendet, um eine Klasse als Request-Handler für RESTful Web Services zu kennzeichnen.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+
+[[bean-scopes]]
+=== Spring-Bean Scopes
+Der Bean-Scope definiert den Lebenszyklus und die Sichtbarkeit einer Bean innerhalb des Containers. Er teilt Spring mit, wie viele Instanzen einer bestimmten Bean erstellt werden sollen und wie lange sie verfügbar sein sollen.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+SpringBoot bietet mehrere Bean-Scopes, darunter:
+
+* Singleton: Eine Singleton-Bean wird nur einmal erstellt, und
+dieselbe Instanz wird für alle Anfragen freigegeben. Sie verbleibt im Kontext der Anwendung, bis die Anwendung heruntergefahren wird. Der Singleton scope ist der Default-Scope in Spring.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+
+* Singleton: Eine Singleton-Bean wird nur einmal erstellt, und
+dieselbe Instanz wird für alle Anfragen freigegeben. Sie verbleibt im Kontext der Anwendung, bis die Anwendung heruntergefahren wird. Der Singleton scope ist der Default-Scope in Spring.
+* Prototyp: Eine neue Instanz der Prototyp-Bean wird jedes Mal erstellt, wenn sie angefordert wird. Sie ermöglicht die Koexistenz mehrerer Instanzen derselben Bean innerhalb der Anwendung.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+dieselbe Instanz wird für alle Anfragen freigegeben. Sie verbleibt im Kontext der Anwendung, bis die Anwendung heruntergefahren wird. Der Singleton scope ist der Default-Scope in Spring.
+* Prototyp: Eine neue Instanz der Prototyp-Bean wird jedes Mal erstellt, wenn sie angefordert wird. Sie ermöglicht die Koexistenz mehrerer Instanzen derselben Bean innerhalb der Anwendung.
+* Request: Für jede HTTP-Anfrage wird eine neue Instanz der request-scoped Bean erstellt. Sie ist nützlich, wenn Sie Daten speichern wollen, die für eine einzelne Anfrage spezifisch sind, wie z.B. Informationen zur Benutzersitzung.
+* Session: Eine neue Instanz der Session-scoped Bean wird für jede Benutzersitzung erstellt. Sie bewahrt den für die Benutzersitzung spezifischen Zustand über mehrere Anfragen hinweg.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+* Prototyp: Eine neue Instanz der Prototyp-Bean wird jedes Mal erstellt, wenn sie angefordert wird. Sie ermöglicht die Koexistenz mehrerer Instanzen derselben Bean innerhalb der Anwendung.
+* Request: Für jede HTTP-Anfrage wird eine neue Instanz der request-scoped Bean erstellt. Sie ist nützlich, wenn Sie Daten speichern wollen, die für eine einzelne Anfrage spezifisch sind, wie z.B. Informationen zur Benutzersitzung.
+* Session: Eine neue Instanz der Session-scoped Bean wird für jede Benutzersitzung erstellt. Sie bewahrt den für die Benutzersitzung spezifischen Zustand über mehrere Anfragen hinweg.
+* Application: Eine neue Instanz der Application-scoped Bean wird einmal für die gesamte Anwendung erstellt. Sie wird für alle Anfragen und Sitzungen gemeinsam genutzt.
```
Nur ein Satz pro Zeile.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -131,52 +74,71 @@ Annotationen, die querschnittliche Aspekte konfigurieren (z.B. `@EnableTransacti
+* WebSocket:  Für jede WebSocket-Sitzung wird eine neue Instanz der WebSocket-scoped Bean erstellt.
+
+[[beans-standardmaessig-als-singletons-definieren]]
+==== Beans standardmäßig als Singletons definieren
```
Hier sollten wir noch die Vorgabe für HTTP-Requests aufnehmen.

dauna-msg commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -259,7 +222,7 @@ Die <<table-vergleich-verfahren-dependency-injection,verschiedenen Verfahren der
+|Kann in Produktiv-Code eingesetzt werden und ist die Empfehlung von Spring.
 
 |Field-Injection
 |Vermeidung von Boilerplate-Code
```
Gemäß Ticket sollte die Empfehlung in einer Vorgabe umgewandelt werden.

stdll committed:
chore: IFS-3028 replace "Spring-Bean" by "Spring Bean"

stdll committed:
docs: IFS-3028 revise page about Spring

- title now includes "Spring Boot"
- descriptions mention backends and batches
- generally forbid Spring AOP for business code, allow it for technical code
- shorten explanations
- rely more on the official documentation of Spring and Spring Boot

stdll committed:
docs: IFS-3028 reinstate deployment options

- servlet containers are still an option in IsyFact 4

stdll committed:
chore: IFS-3028 fix wording issue

stdll committed:
chore: IFS-3028 add changelog entry

mtpham-collab commented on isyfact-standards-doc/src/docs/antora/modules/einstieg/pages/tutorial/inhalt.adoc:
```diff
@@ -185,11 +185,11 @@ a| Interfaces zur Definition der Schnittstelle der Komponente "Auskunft". +
-|*AuskunftImpl* |Implementierung der Komponente `Auskunft`. Diese Klasse wird als Spring-Bean konfiguriert.
-Weitere benötigte Komponenten (Spring-Beans) werden dieser Komponente per Spring-Dependency-Injection bekannt gemacht.
+|*AuskunftImpl* |Implementierung der Komponente `Auskunft`. Diese Klasse wird als Spring Bean konfiguriert.
+Weitere benötigte Komponenten (Spring Beans) werden dieser Komponente per Spring-Dependency-Injection bekannt gemacht.
```
```suggestion
Weitere benötigte Komponenten (Spring Beans) werden dieser Komponente per Spring Dependency Injection bekannt gemacht.
```

mtpham-collab commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -1,310 +1,188 @@
-Das Spring-Framework ist ein Java EE Framework, welches im Kern sehr verständlich und leicht zu verwenden ist.
-In ihm werden die Bestandteile eines Systems als _Spring-Beans_ definiert.
-Neben seiner Kern-Funktionalität der Verwaltung, Konfiguration und aspektorientierten Erweiterung von Beans bietet Spring viele Funktionalitäten, welche die Entwicklung einer Anwendung erleichtern.
+Spring und das auf ihm basierende Spring Boot sind leichtgewichtige, Java-basierte Anwendungsframeworks.
```
```suggestion
Spring und das darauf basierende Spring Boot sind leichtgewichtige, Java-basierte Anwendungsframeworks.
```

mtpham-collab commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -1,310 +1,188 @@
 
-Die folgenden Inhalte beschreiben die Nutzung von Spring für die allgemeine Umsetzung von Backends.
-Die Nutzung spezifischer Features von Spring sind Bestandteile der folgenden Bausteine:
+_Spring Beans_ sind Java-Objekte (POJOs), die durch den https://docs.spring.io/spring-framework/reference/core/beans.html[IoC Container] von Spring instanziiert und verwaltet werden.
```
```suggestion
_Spring Beans_ sind Java-Objekte (POJOs), die durch den https://docs.spring.io/spring-framework/reference/core/beans.html[IoC Container] von Spring instanziiert, konfiguriert und verwaltet werden.
```

mtpham-collab commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -1,310 +1,188 @@
 
-[[listing-applikationsklasse-mit-springbootapplication-annotation]]
-.Applikationsklasse mit `@SpringBootApplication`
+Die Annotation `@SpringBootApplication` markiert eine Konfigurationsklasse, die eine oder mehrere `@Bean`-Methoden deklariert und außerdem die automatische Konfiguration und das Scannen von Komponenten auslöst.
```
```suggestion
Die Annotation `@SpringBootApplication` markiert eine Konfigurationsklasse, die eine oder mehrere `@Bean`-Methoden deklariert.
Darüber hinaus löst sie die automatische Konfiguration sowie das Scannen von Komponenten aus.
```

mtpham-collab commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -1,310 +1,188 @@
+
+=== Konfiguration von Schichten
+
+Für jede xref:software-technisch/backend.adoc#systemarchitektur-backend[Schicht] wird im Package der Schicht (`<org>.<domäne>.<anwendung>.<it-system>.<schicht>`) parallel zu den Interfaces der Komponentenschnittstellen eine mit `@Configuration` annotierte <<listing-konfigurationsklasse,Konfigurationsklasse>> erstellt.
```
```suggestion
Für jede xref:software-technisch/backend.adoc#systemarchitektur-backend[Schicht] wird im entsprechenden Package (`<org>.<domäne>.<anwendung>.<it-system>.<schicht>`) parallel zu den Interfaces der Komponentenschnittstellen eine mit `@Configuration` annotierte <<listing-konfigurationsklasse,Konfigurationsklasse>> erstellt.
```

mtpham-collab commented on isyfact-standards-doc/src/docs/antora/modules/referenzarchitektur/pages/software-technisch/backend/spring.adoc:
```diff
@@ -1,310 +1,188 @@
-Für diese Aspekte werden im folgenden Vorgaben aufgestellt.
+[[konfiguration-von-spring-beans]]
+=== Konfiguration von Komponenten
+Die Spring Beans der Komponenten werden mit der Annotation `@Component` bzw. deren Spezialisierungen (`@Repository`, `@Service`, `@Controller` etc.) versehen.
```
```suggestion
Die Spring Beans der Komponenten werden mit der Annotation `@Component` oder deren Spezialisierungen wie `@Repository`, `@Service`, `@Controller` und weiteren, je nach Einsatzzweck, versehen.
```

stdll committed:
docs: IFS-3028 apply suggestions from code review

Co-authored-by: mtpham-collab <74294456+mtpham-collab@users.noreply.github.com>

stdll committed:
Merge branch 'master' into feature/IFS-3028-Lockerung-der-Vorgaben-zurNutzung-von-Spring
